README.md
# micro-rpc-client

[![Build Status](https://travis-ci.org/bufferapp/micro-rpc-client.svg?branch=master)](https://travis-ci.org/bufferapp/micro-rpc-client)

Universal JS RPC Client https://github.com/bufferapp/buffer-rpc/

## Quickstart

Create an RPC Client and call an add function

```js
const RPCClient = require('micro-rpc-client');

const client = new RPCClient({
  url: 'https://localhost:3000/rpc',
});

const main = async () => {
  const result = await client.call('add', [1, 2]);
  console.log(result); // 3
};
main();
```

## Usage

A few examples of how to call client methods

```js
const RPCClient = require('micro-rpc-client');

const client = new RPCClient({
  url: 'https://localhost:3000/rpc',
  sendCredentials: 'same-origin', // send cookies on same origin requests
});

const main = async () => {
  const methods = await client.listMethods();
  console.log(method);
  /*
  [
    {
      "docs": "add two numbers"
      "name": "add"
    },
    {
      "docs": "list all available methods",
      "name": "methods"
    }
  ]
  */
  const result = await client.call('add', [1, 2]);
  console.log(result); // 3
};
main();
```

## API

### constructor

Create an instance of RPCClient

```js
const client = new RPCClient({
  url,
  sendCredentials,
});
```

**url** - _string_ - full url to the RPC Endpoint Server  
**sendCredentials** - _string_ - when to send things like cookies with a request, passes arguments to the credentials init argument in the fetch API https://developer.mozilla.org/en-US/docs/Web/API/WindowOrWorkerGlobalScope/fetch#Parameters

Returns an instance of the RPC client

### listMethods

List all remote methods on a server

```js
const methods = await client.listMethods()
```

Returns a promise that resolves with a list of the remote methods

### call

```js
const result = await client.call(name, args)
```

**name** - _string_ - name of the remote RPC endpoint  
**args** - _array_ or _object_ - arguments to pass to the remote RPC endpoint

Returns a promise that resolves with the result and rejects with an error.

## Error Handling

When calling a remote function with `call` there are different categories of responses:

### Success

This is the typical success case

```
statusCode = 200
result = {} // JSON
```

### Fail (Handled)

An error triggered from [createError](https://github.com/bufferapp/buffer-rpc/blob/master/README.md#createerror):

```
statusCode = 400
result = {
  error: 'string',
  code: 1000, // or some custom code
  handled: true,
}
```

### Fail (Unhandled)

An error triggered from [errorMiddleware](https://github.com/bufferapp/buffer-rpc/blob/master/README.md#createerror):

```
statusCode = 500
result = {
  error: 'string',
  code: 1000, // or some custom code
  handled: true,
}
```

### Fail (Unexpected)

If the [errorMiddleware](https://github.com/bufferapp/buffer-rpc/blob/master/README.md#createerror) is not hooked up, or something unexpected happened the original error will be thrown.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.4.1]
### Added
- Added support for passing a headers object when calling RPC endpoints
### Note
- Sorry about the huge version bump

## [1.0.0]
### Changed
- [Breaking] The RPC client now makes calls to an RPC server in the form of `endpoint/:methodName`, to allow for clearer debugging and tracing. Ensure you are using >= v. 1.0.0 of `buffer-rpc` on your server as well.
### Added
- Added this CHANGELOG.

index.test.js
const fetch = require('isomorphic-fetch');
const RPCClient = require('./index');

describe('RPCClient', () => {
  it('should create an RPCClient', () => {
    expect(new RPCClient()).toBeDefined();
  });

  it('should set RPC server url', () => {
    const url = 'http://some-rpc-server';
    const rpc = new RPCClient({ url });
    expect(rpc.url).toBe(url);
  });

  it('should set default RPC server url', () => {
    const rpc = new RPCClient();
    expect(rpc.url).toBe('http://localhost');
  });

  describe('listMethods', () => {
    it('should list available RPC methods', () => {
      const rpc = new RPCClient();
      return rpc.listMethods().then((methods) => {
        expect(fetch).toBeCalledWith('http://localhost/methods', {
          method: 'POST',
          headers: {
            Accept: 'application/json',
            'Content-Type': 'application/json',
          },
          body: JSON.stringify({}),
        });
        expect(methods).toEqual(fetch.fakeMethods);
      });
    });
  });

  describe('call', () => {
    it('should call RPC method', () => {
      const name = 'someMethod';
      const rpc = new RPCClient();
      return rpc.call(name).then((response) => {
        expect(fetch).toBeCalledWith('http://localhost/someMethod', {
          method: 'POST',
          headers: {
            Accept: 'application/json',
            'Content-Type': 'application/json',
          },
          body: JSON.stringify({}),
        });
        expect(response).toEqual(fetch.fakeResponse);
      });
    });

    it('should call RPC method with args', () => {
      const name = 'someMethod';
      const args = { a: 'a', b: 'b' };
      const rpc = new RPCClient();
      return rpc.call(name, args).then((response) => {
        expect(fetch).toBeCalledWith('http://localhost/someMethod', {
          method: 'POST',
          headers: {
            Accept: 'application/json',
            'Content-Type': 'application/json',
          },
          body: JSON.stringify({
            args: JSON.stringify(args),
          }),
        });
        expect(response).toEqual(fetch.fakeResponse);
      });
    });

    it('should trigger error when response !== 200', async () => {
      expect.assertions(4);
      const name = 'shouldThrow';
      const rpc = new RPCClient();
      try {
        await rpc.call(name);
      } catch (err) {
        expect(err.handled).toBe(true);
        expect(err.message).toBe('this method should throw an error');
        expect(err.code).toBeUndefined();
        expect(err.status).toBe(400);
      }
    });

    it('should trigger error when response !== 200 with custom code', async () => {
      expect.assertions(5);
      const name = 'shouldThrowCustomCode';
      const rpc = new RPCClient();
      try {
        await rpc.call(name);
      } catch (err) {
        expect(err.message).toBe('this method should throw an error');
        expect(err.code).toBeDefined();
        expect(err.code).toBe(fetch.fakeCode);
        expect(err.handled).toBe(true);
        expect(err.status).toBe(400);
      }
    });

    it('should trigger an unhandled error when response === 500', async () => {
      expect.assertions(3);
      const name = 'shouldThrowJSON500';
      const rpc = new RPCClient();
      try {
        await rpc.call(name);
      } catch (err) {
        expect(err.message).toBe('some unexpected error occured');
        expect(err.handled).toBe(false);
        expect(err.status).toBe(500);
      }
    });

    it('should reject when json parse fails', async () => {
      expect.assertions(1);
      const name = 'shouldFailJson';
      const rpc = new RPCClient();
      try {
        await rpc.call(name);
      } catch (err) {
        expect(err.message).toBe('something went wrong parsing json');
      }
    });
  });

  describe('headers', () => {
    it('should call RPC method with headers', async () => {
      const name = 'someMethod';
      const args = { a: 'a', b: 'b' };
      const headers = { foo: 'bar' };
      const rpc = new RPCClient();
      const response = await rpc.call(name, args, headers);

      expect(fetch).toBeCalledWith('http://localhost/someMethod', {
        method: 'POST',
        headers: {
          Accept: 'application/json',
          'Content-Type': 'application/json',
          ...headers,
        },
        body: JSON.stringify({
          args: JSON.stringify(args),
        }),
      });
      expect(response).toEqual(fetch.fakeResponse);
    });
  });
});


index.js
const fetch = require('isomorphic-fetch');

class RPCClient {
  constructor(options = {}) {
    this.url = options.url || 'http://localhost';
    this.sendCredentials = options.sendCredentials;
  }

  listMethods() {
    return this.call('methods');
  }

  call(name, args, headers = {}) {
    return fetch(`${this.url}/${name}`, {
      method: 'POST',
      headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json',
        ...headers,
      },
      body: JSON.stringify({
        args: JSON.stringify(args),
      }),
      credentials: this.sendCredentials,
    })
      .then(
        response =>
          new Promise((resolve, reject) => {
            response
              .json()
              .then(parsedResponse =>
                resolve({
                  response: parsedResponse,
                  status: response.status,
                }),
              )
              .catch(error => reject(error));
          }),
      )
      .then(({ response, status }) => {
        if (response.error) {
          const err = new Error(response.error);
          err.code = response.code;
          err.handled = response.handled;
          err.status = status;
          throw err;
        }
        return response;
      })
      .then(response => response.result);
  }
}

module.exports = RPCClient;


__mocks__/isomorphic-fetch.js
const fakeMethods = ['methods', 'test'];
const fakeResponse = 'fake response';
const fakeCode = 1001;

const parseMethodFromUrl = (url) => {
  const split = url.split('/');
  return split.pop();
};

const fetch = jest.fn((url) => {
  if (parseMethodFromUrl(url) === 'methods') {
    return Promise.resolve({
      status: 200,
      json: () =>
        Promise.resolve({
          result: fakeMethods,
        }),
    });
  } else if (parseMethodFromUrl(url) === 'shouldThrow') {
    return Promise.resolve({
      status: 400,
      json: () =>
        Promise.resolve({
          error: 'this method should throw an error',
          handled: true,
        }),
    });
  } else if (parseMethodFromUrl(url) === 'shouldThrowCustomCode') {
    return Promise.resolve({
      status: 400,
      json: () =>
        Promise.resolve({
          error: 'this method should throw an error',
          code: fakeCode,
          handled: true,
        }),
    });
  } else if (parseMethodFromUrl(url) === 'shouldThrowJSON500') {
    return Promise.resolve({
      status: 500,
      json: () =>
        Promise.resolve({
          error: 'some unexpected error occured',
          handled: false,
        }),
    });
  } else if (parseMethodFromUrl(url) === 'shouldFailJson') {
    return Promise.resolve({
      status: 500,
      json: () => Promise.reject(new Error('something went wrong parsing json')),
    });
  }
  return Promise.resolve({
    status: 200,
    json: () =>
      Promise.resolve({
        result: fakeResponse,
      }),
  });
});
fetch.fakeMethods = fakeMethods;
fetch.fakeResponse = fakeResponse;
fetch.fakeCode = fakeCode;
module.exports = fetch;
